src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder.cs

```diff
@@ -25,573 +65,740 @@ namespace Microsoft.CodeAnalysis.FindSymbols
+
+            var currentMetadataTypes = new HashSet<INamedTypeSymbol>(SymbolEquivalenceComparer.Instance);
+            var currentSourceAndMetadataTypes = new HashSet<INamedTypeSymbol>(SymbolEquivalenceComparer.Instance);
 
-            var projectTasks = new List<Task>();
-            foreach (var project in dependentProjects)
+            currentSourceAndMetadataTypes.Add(type);
+            if (searchInMetadata)
             {
-                projectTasks.Add(Task.Run(
-                    async () => await GetDependentTypesInProjectAsync(type, project, solution, predicate, cache, locationsInMetadata, results, cancellationToken).ConfigureAwait(false), cancellationToken));
+                currentMetadataTypes.Add(type);
             }
 
-            await Task.WhenAll(projectTasks).ConfigureAwait(false);
+            var result = new HashSet<INamedTypeSymbol>(SymbolEquivalenceComparer.Instance);
 
-            if (results.Any())
+            // There is a data dependency when searching projects.  Namely, the derived types we
+            // find in one project will be used when searching successive projects.  So we cannot
+            // searchthese projects in parallel.
+            foreach (var project in orderedProjectsToExamine)
             {
-                return results.OfType<INamedTypeSymbol>();
-            }
-            else
-            {
-                return SpecializedCollections.EmptyEnumerable<INamedTypeSymbol>();
+                await findTypesInProjectAsync(
+                    searchInMetadata, result,
+                    currentMetadataTypes, currentSourceAndMetadataTypes,
+                    project, transitive, cancellationToken).ConfigureAwait(false);
             }
+
+            return result;
         }
 
-        private static async Task GetDependentTypesInProjectAsync(
-            INamedTypeSymbol type, Project project, Solution solution, Func<INamedTypeSymbol, INamedTypeSymbol, bool> predicate, ConditionalWeakTable<Compilation, ConcurrentDictionary<SymbolKey, List<SymbolKey>>> cache, bool locationsInMetadata, ConcurrentSet<ISymbol> results, CancellationToken cancellationToken)
+        private static Task FindDerivedClassesInProjectAsync(
+            bool searchInMetadata,
+            HashSet<INamedTypeSymbol> result,
+            HashSet<INamedTypeSymbol> currentMetadataTypes,
+            HashSet<INamedTypeSymbol> currentSourceAndMetadataTypes,
+            Project project,
+            bool transitive,
+            CancellationToken cancellationToken)
         {
-            cancellationToken.ThrowIfCancellationRequested();
-
-            var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+            return FindTypesInProjectAsync(searchInMetadata, result,
+                currentMetadataTypes, currentSourceAndMetadataTypes,
+                project,
+                findMetadataTypesAsync: FindDerivedMetadataClassesInProjectAsync,
```

Comment:
probably.  I'm also likely going to refactor this a bit.  Ther'es a very diamond-ish pattern from these calls.  The varying ones call a common one which calls out to specific methods which call a common method etc. etc.

I think i can just have the topmost method collect the varying behavior it wants, and then pass that immediately to a common core which just keeps calling common core methods.